src/base/format/float_node.h

```diff
@@ -22,31 +22,97 @@
 #include <cstring>
 
 #include "base/format/abstract_node.h"
+#include "base/format/write_number.h"
 
 namespace format_impl {
 
+// We assume that it is a single displayed character, but we allow it to be
+// stored as a multi-byte UTF-8 sequence.
+inline char* write_decimal_separator(char* out, bool localize) {
+	if (localize) {
+		for (const char* dec_sep = pgettext("decimal_separator", "."); *dec_sep; ++dec_sep) {
+			*out = *dec_sep;
+			++out;
+		}
+	} else {
+		*out = '.';
+		++out;
+	}
+	return out;
+}
+
 struct FloatNode : FormatNode {
-	FloatNode(const uint8_t f, const size_t w, const int32_t p) : FormatNode(f, w, p) {
+	FloatNode(const uint8_t f, const size_t w, const int32_t p, bool dyn_prec)
+	   : FormatNode(f, w, p), dynamic_precision_(dyn_prec) {
+		rounding_ = 0.5;
+		precision_multiplier_ = 1.;
+		for (unsigned i = precision_; i > 0; --i) {
+			rounding_ /= 10.;
+			precision_multiplier_ *= 10.;
+		}
+		check_zero_extra_precision_ = kDynamicPrecisionExtra / precision_multiplier_;
 	}
 
 	char* append(char* out, const ArgType t, Argument arg_u, const bool localize) const override {
-		double arg;
+		int64_t int_part;
+		int64_t fractional = 0;
+		bool is_negative = false;
+		bool is_zero = false;
+		size_t current_precision = precision_;
+
 		switch (t) {
 		case ArgType::kFloat:
-			if (arg_u.float_val > kMaxInt || arg_u.float_val < -kMaxInt) {
+			double rounded;
+			double frac;
```

Comment:
Same here, `switch` `case`s declaring a local variable should always be inside `{}`